aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -583,6 +615,150 @@ impl<T: Transaction> CapturedReads<T> {
         })
     }
 
+    /// Records the read to [CrossBlockModuleCache].
+    pub(crate) fn capture_global_cache_read(&mut self, module_id: ModuleId) {
+        self.module_reads.insert(module_id, ModuleRead::GlobalCache);
+    }
+
+    /// Records the read to [SyncCodeCache].
+    pub(crate) fn capture_block_cache_read(
+        &mut self,
+        module_id: ModuleId,
+        read: Option<Arc<MaybeCommitted<ModuleCacheEntry>>>,
+    ) {
+        self.module_reads
+            .insert(module_id, ModuleRead::BlockCache(read));
+    }
+
+    /// If the module has been previously read from [SyncCodeCache], allows to query information
+    /// about it.
+    fn fetch_from_module_read<F, V>(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+        map_func: F,
+    ) -> CacheRead<V>
+    where
+        F: Fn(Option<&Arc<MaybeCommitted<ModuleCacheEntry>>>) -> V,
+    {
+        match self.module_reads.get(&(address, module_name)) {
+            Some(ModuleRead::BlockCache(read)) => CacheRead::Hit(map_func(read.as_ref())),
+            None => CacheRead::Miss,
+            Some(ModuleRead::GlobalCache) => unreachable!("Global cache reads are not cached"),
```

Comment:
This is an invariant due to the use of CapturedReads, not a local invariant, right? (since it can capture GlobalCache read) - so we should probably not panic and return an error that can fallback as well.

Reply:
Indeed. Changed to panic errors, and simplified captured reads